lib/ddtrace/workers/async.rb

```diff
@@ -0,0 +1,163 @@
+require 'ddtrace/logger'
+
+module Datadog
+  module Workers
+    module Async
+      # Adds threading behavior to workers
+      # to run tasks asynchronously.
+      # rubocop:disable Metrics/ModuleLength
+      module Thread
+        FORK_POLICY_STOP = :stop
+        FORK_POLICY_RESTART = :restart
+        SHUTDOWN_TIMEOUT = 1
+
+        def self.included(base)
+          base.send(:prepend, PrependedMethods)
+        end
+
+        # Methods that must be prepended
+        module PrependedMethods
+          def perform(*args)
+            start { self.result = super(*args) } if unstarted?
```

Comment:
Only consideration for this is that we'd have to invert the logic that drives this: `worker.nil? || forked?`. Does `!worker.nil? && !forked?` strictly mean the same thing? I don't think it does.

Some context: when the thread is running, and you fork the process, within the fork the worker thread will be present but dead. So the question is should you restart a dead fork? This is what the `fork_policy` variable drives.

Since it is initially dead in the fork, `unstarted?` is supposed to allow the fork to start the worker as if its never been started. Issue is if you adopt `!worker.nil? && !forked?` as `started?` it will always return `true` because the dead worker thread will be hanging around since the beginning of the fork, thus it would never attempt to start the thread in the fork. You also can't tell if a worker thread was never started in the fork or if it started and failed just by looking at the state of `worker`.

I'm inclined to keep this as is, unless there's a cleaner, logical way to handle this without making forking behavior more difficult to implement.

Reply:
Your boolean negation seems correct to me: `!worker.nil? && !forked?`.

>  Issue is if you adopt !worker.nil? && !forked? as started? it will always return true because the dead worker thread will be hanging around since the beginning of the fork, thus it would never attempt to start the thread in the fork.


`!forked?` will return false after the fork, making `!worker.nil? && !forked?` return `false` in this case.

If you are not comfortable with the new expression, we can add a `!` to the beginning to the method, then there's no ambiguity:
```ruby
def started?
  !(worker.nil? || forked?)
end
```

Reply:
Both statements are functionally equivalent, but you are right about `!forked?` returning `false`.

I already have tests written for this forking scenario, so I should be able to just try this and see if it works or not.

Reply:
Made this change.